apps/daimo-mobile/src/logic/systemContacts.ts

```diff
@@ -0,0 +1,149 @@
+
+    if (!contactsPermission.granted) {
+      await requestContactsAccess(contactsPermission.canAskAgain);
+    }
+    Contacts.getPermissionsAsync().then(setContactsPermission);
+  };
+
+  return { permission: contactsPermission, ask };
+}
+
+// Take user through a two-step flow to ask for contacts access.
+// Promise resolves when user has finished the flow (either granting
+// or denying).
+function requestContactsAccess(canAskAgain: boolean) {
+  // Step 1: Ask with our own prompt, re-assuring user.
+  return new Promise<void>((resolve) => {
+    Alert.alert(
+      "Daimo Would Like to Access Your Contacts",
+      "Your contacts are accessed locally to find and pay friends with Daimo.\n\nThey are never uploaded to our servers.",
+      [
+        {
+          text: `Continue`,
+          onPress: () => {
+            if (canAskAgain) askSystem(resolve);
+            else askOpenSettings(resolve);
+          },
+        },
+        {
+          text: "Don't Allow",
+          style: "cancel",
+          onPress: () => {
+            resolve();
+          },
+        },
+      ]
+    );
+  });
+}
+
+function askSystem(resolve: (value: void) => void) {
+  // Step 2: If user says yes to our own prompt, ask with the system prompt
+  Contacts.requestPermissionsAsync().then(() => {
+    resolve();
+  });
+}
+
+async function askOpenSettings(resolve: (value: void) => void) {
+  // Step 2: If user says yes to our own prompt but previously denied system
+  // prompt, ask them to open settings.
+  Alert.alert(
+    "Enable access in Settings",
+    "Visit Settings > Daimo and enable contacts.",
+    [
+      {
+        text: "Continue",
+        onPress: () => {
+          Linking.openSettings().then(async () => {
+            // On iOS, the app is reset by the OS when access is changed from
+            // settings, so this isn't triggered. On Android, we wait a
+            // few seconds to let user open settings, then resolve.
+            await new Promise((f) => setTimeout(f, 5000));
+            resolve();
+          });
+        },
+      },
+      {
+        text: "Cancel",
+        style: "cancel",
+        onPress: () => {
+          resolve();
+        },
+      },
+    ]
+  );
+}
+
+export function useSystemContactsSearch(
+  prefix: string,
+  enabled: boolean
+): SystemContact[] {
+  const [recipients, setRecipients] = useState<SystemContact[]>([]);
+
+  useEffect(() => {
+    const search = async (nameSearchTerm: string) => {
+      const { data } = await Contacts.getContactsAsync({
+        // TODO: add images?
+        fields: [Contacts.Fields.PhoneNumbers, Contacts.Fields.Emails],
+        name: nameSearchTerm,
+        pageSize: 5,
+      });
+
+      const matches = [] as SystemContact[];
+      for (const contact of data) {
+        const name = contact.name;
+        if (name == null || !name.toLowerCase().includes(nameSearchTerm))
+          continue;
+
+        // Show atmost one email and one phone number per contact.
```

Comment:
nit: at most

Reply:
below, aren't we adding *all* phone numbers and all emails?

Reply:
there's a `break` after the first hit?

Reply:
done nit